src/xmlparser.rs

```diff
@@ -0,0 +1,423 @@
+    error_emitter: Rc<ErrorEmitter>,
+}
+
+impl Element {
+    /// Returns the element local name.
+    pub fn name(&self) -> &str {
+        &self.name.local_name
+    }
+
+    /// Value of attribute with given name or None if it is not found.
+    pub fn attr(&self, name: &str) -> Option<&str> {
+        for attr in &self.attributes {
+            if attr.name.local_name == name {
+                return Some(&attr.value);
+            }
+        }
+        None
+    }
+
+    /// Checks if elements has any attributes.
+    pub fn has_attrs(&self) -> bool {
+        !self.attributes.is_empty()
+    }
+
+    pub fn attr_bool(&self, name: &str, default: bool) -> bool {
+        for attr in &self.attributes {
+            if attr.name.local_name == name {
+                return attr.value == "1";
+            }
+        }
+        default
+    }
+
+    /// Returns element position.
+    pub fn position(&self) -> TextPosition {
+        self.position
+    }
+
+    /// Value of attribute with given name or an error when absent.
+    pub fn attr_required(&self, name: &str) -> Result<&str> {
+        for attr in &self.attributes {
+            if attr.name.local_name == name {
+                return Ok(&attr.value);
+            }
+        }
+        let message = format!("Attribute `{}` on element <{}> is required.", name, self.name());
+        Err(self.error_emitter.emit(&message, self.position))
+    }
+}
+
+impl<'a> XmlParser<'a> {
+    pub fn from_path(path: &Path) -> Result<XmlParser> {
+        let file = File::open(&path)
+            .chain_err(|| format!("Can't open file: {}", path.display()))?;
+        Ok(XmlParser {
+            parser: EventReader::new(Box::new(BufReader::new(file))),
+            peek_event: None,
+            peek_position: TextPosition::new(),
+            error_emitter: Rc::new(ErrorEmitter {
+                path: Some(path.to_owned()),
+            }),
+        })
+    }
+
+    #[cfg(test)]
+    pub fn new<'r, R: 'r + Read>(read: R) -> Result<XmlParser<'r>> {
+        Ok(XmlParser {
+            parser: EventReader::new(Box::new(read)),
+            peek_event: None,
+            peek_position: TextPosition::new(),
+            error_emitter: Rc::new(ErrorEmitter {
+                path: None
+            }),
+        })
+    }
+
+    /// Returns an error that combines current position and given error message.
+    pub fn fail(&self, message: &str) -> Error {
+        self.error_emitter.emit(message, self.position())
+    }
+
+    /// Returns an error that combines given error message and position.
+    pub fn fail_with_position(&self, message: &str, position: TextPosition) -> Error {
+        self.error_emitter.emit(message, position)
+    }
+
+    pub fn unexpected_element(&self, elem: &Element) -> Error {
+        let message = format!("Unexpected element <{}>", elem.name());
+        self.error_emitter.emit(&message, elem.position())
+    }
+
+    pub fn unexpected_event(&self, event: &XmlEvent) -> Error {
```

Comment:
As no `XmlEvent` now used outside, maybe make it private?

Reply:
*not used

Reply:
Good point. Done.